File: algorithms/1_easy/D/01704_determine_if_string_halves_are_alike/solution.py
class Solution:
    def halvesAreAlike(self, s: str) -> bool:
        s_len = len(s)
        half =  s_len / 2
        balance = 0

        for i in range(s_len):
            match s[i]:
                case 'a' | 'e' | 'i' | 'o' | 'u' | 'A' | 'E' | 'I' | 'O' | 'U':
                    if i < half:
                        balance += 1
                    else:
                        balance -= 1

        return balance == 0

File: algorithms/1_easy/D/01704_determine_if_string_halves_are_alike/solution.cpp
class Solution {
public:
    bool halvesAreAlike(string s) {

        const size_t len{s.size()};
        size_t half{len / 2};
        int balance{0};

        for (size_t i{0}; i < len; ++i) {
            switch (s[i]) {
                case 'a':
                case 'e':
                case 'i':
                case 'o':
                case 'u':
                case 'A':
                case 'E':
                case 'I':
                case 'O':
                case 'U':
                    if (i < half) {
                        balance += 1;
                    } else {
                        balance -= 1; 
                    }
                    break;
                default:
                    break;
            }
        }

        return balance == 0;
    }
};
